### PR TITLE
ADBDEV-1885. Add `--compression-type`. Refactor compression in `helper`. Implement `zstd` compression

### PR DESCRIPTION
**Pull request by leskin-in**
Add a flag to set compression type. `gzip` is the only currently supported value. To disable compression, `--no-compression` or `--compression-level 0` are to be set. `gzip` is also the default value for this parameter.

In `helper`, introduce `backup_helper_pipes.go`. This defines wrappers for buffer writers encapsulating various compression types (currently "no compression" and "gzip"). This allows to drop (somewhat) complex `Writer` closure algorithm in `backup_helper.go`.

Update `helper_test` and `compression_test` according to new changes.

Add support for `zstd` compression type. Update tests and run option descriptions accordingly.

### CODE REVIEW

**Comment by leskin-in** [comment]
@Stolb27 please add other reviewers whose contribution you think could be useful.

**Comment by hughcapet** [comment]
I suppose we should add the same validation for the new flag: --no-compression and --compression-type are exclusive
https://github.com/arenadata/gpbackup/blob/c06438097db52307acab3681a3dbba829f49044e/backup/validate.go#L102

**Comment by leskin-in** [comment]
Thanks @hughcapet . Fixed in https://github.com/arenadata/gpbackup/pull/2/commits/a34b75eb0d68606b4b8be02a352bd97a2f8f04bb. In this commit, I have also modified the compression type & level validation function, introducing a more flexible implementation.

File: backup/backup.go [hunk 1]
```diff
@@ -268,7 +268,7 @@ func backupData(tables []Table) {
 		}
 		utils.WriteOidListToSegments(oidList, globalCluster, globalFPInfo)
 		utils.CreateFirstSegmentPipeOnAllHosts(oidList[0], globalCluster, globalFPInfo)
-		compressStr := fmt.Sprintf(" --compression-level %d", MustGetFlagInt(options.COMPRESSION_LEVEL))
+		compressStr := fmt.Sprintf(" --compression-level %d --compression-type %s", MustGetFlagInt(options.COMPRESSION_LEVEL), MustGetFlagString((options.COMPRESSION_TYPE)))
```
Comment (hughcapet) [inline backup/backup.go:271]:
```suggestion
		compressStr := fmt.Sprintf(" --compression-level %d --compression-type %s", MustGetFlagInt(options.COMPRESSION_LEVEL), MustGetFlagString(options.COMPRESSION_TYPE))
```

**Comment by Stolb27** [comment]
@leskin-in
Do we really need unreleased yet changes from upstream master? These commits aren't included in any tag. But may contain unstable changes.
Please change base (and rebase previously) branch to origin/master if not.

**Comment by leskin-in** [comment]
@Stolb27 ,
we are going to open a PR to upstream, and then backport changes into our `master` branch. For this reason, there is no use in branching from our `master` or some older version (with tag) of `upstream`'s `master`.

**Review by hughcapet — APPROVED** [review]

**Review by Stolb27 — CHANGES_REQUESTED** [review]
@leskin-in, there is at least one unimplemented moment in this patch. It is the case of restoration from a backup created by a previous version. There is an appropriate end_to_end test in the test suite. On this branch, this test hangs. gpdb copy command wait for data from the pipe, but nobody writes to this pipe because helper failed with: 

```
20210729:01:00:42 gpbackup_helper:gpadmin:08ea34454fe9:013219-[ERROR]:-open /tmp/temp236231718/corrupt-db/demoDataDir2/backups/20190809/20190809230424/gpbackup_2_20190809230424: no such file or directory: goroutine 1 [running]:
runtime/debug.Stack(0x8c01c0, 0xc00018e2a0, 0xc00006e840)
        /usr/lib/golang/src/runtime/debug/stack.go:24 +0x9f
github.com/greenplum-db/gpbackup/helper.DoHelper()
        /home/gpadmin/go/src/github.com/greenplum-db/gpbackup/helper/helper.go:90 +0x168
main.main()
        /home/gpadmin/go/src/github.com/greenplum-db/gpbackup/gpbackup_helper.go:10 +0x25
20210729:01:00:42 gpbackup_helper:gpadmin:08ea34454fe9:013219-[DEBUG]:-Segment 2: Cleanup complete
```

But backup created with gzip: 

```
[root@08ea34454fe9 gpadmin]# ls /tmp/temp236231718/corrupt-db/demoDataDir2/backups/20190809/20190809230424/
gpbackup_2_20190809230424.gz  gpbackup_2_20190809230424_toc.yaml
```

This backup is compressed but compression type obviously omitted:

```
backupdir: /tmp/corrupt-db
backupversion: 1.12.1
compressed: true
databasename: '"corrupt-db"'
databaseversion: 4.3.33.2 build 1
...
```

Please fix up end_to_end pipeline. All tests passed successfully on the base branch.
I've describe test running process in https://arenadata.atlassian.net/browse/ADBDEV-1918

**Comment by leskin-in** [comment]
@Stolb27 ,
the mentioned problem is fixed in https://github.com/arenadata/gpbackup/pull/2/commits/80bb1c4e90515f571117c0a1ff0862d1851504c4.

The change effectively means `gprestore` is backward-compatible.

**Review by Stolb27 — APPROVED** [review]

**Review by maksm90 — APPROVED** [review]

**Comment by Stolb27** [comment]
I think we may close this one. Further review is moving to upstream #523

**Pull request closed without merging**